File: src/taxoquiz/explore.py
"""Free browsing of the tree, with no game attached.

Explore mode has no secret and no guesses: you start anywhere and walk the tree.
That makes it a different data problem from the game. `game_state` returns the
union of a handful of lineages — tens of nodes — whereas explore can be pointed
at a root with 27,000 descendants.

So the tree is served in **slices**: a request names a root and a depth, and
nodes with children beyond that depth come back marked `truncated`, for the
client to fetch when the user opens them. `depth=None` returns everything, which
is supported deliberately (it is the honest answer to "what if I just render the
whole thing?") but is not what the UI asks for by default — see the note on
node counts in the API.

Every node carries `species_count`, the number of leaves beneath it. Browsing a
taxonomy without it is guesswork: "Aves" and "Onychophora" look identical as
labels, and one holds a thousand species while the other holds three.
"""

from .game.tree import load_tree

_tree: dict | None = None
_by_name: dict[str, dict] = {}
_parent: dict[str, str] = {}
_depth: dict[str, int] = {}
_species_count: dict[str, int] = {}
_node_count: dict[str, int] = {}


def _ensure_index() -> None:
    """Build the name-keyed indexes once.

    Names are unique across the tree — `datagen/extract_game_tree.py` enforces
    that with `make_names_unique` and refuses to write a tree where it fails —
    so a name is a usable key here. If that ever stopped holding, this index
    would silently collapse distinct taxa onto one entry, which is why the
    guarantee lives in the converter rather than being assumed here.
    """
    global _tree
    if _tree is not None:
        return
    _tree = load_tree()

    def walk(node: dict, parent: str | None, depth: int) -> int:
        name = node["name"]
        _by_name[name] = node
        _depth[name] = depth
        if parent is not None:
            _parent[name] = parent
        children = node.get("children") or []
        if not children:
            _species_count[name] = 1
            _node_count[name] = 1
            return 1, 1
        species = nodes = 0
        for child in children:
            s, n = walk(child, name, depth + 1)
            species += s
            nodes += n
        _species_count[name] = species
        _node_count[name] = nodes + 1
        return species, nodes + 1

    walk(_tree, None, 0)


def _select(node: dict, depth: int | None, budget: int | None) -> set[str]:
    """Names to include, grown breadth-first until a limit bites.

    A **node budget** rather than a depth limit, because depth is the wrong knob
    on a real taxonomy. The Wikidata tree opens with a single-child chain —
    Animalia › Eumetazoa › ParaHoxozoa › Bilateria › Nephrozoa › … — so three
    levels down from the root is nine nodes and five clicks from anything
    interesting, while three levels down from a bushy genus is hundreds.

    Growing level by level and stopping before the budget is exceeded adapts to
    whichever shape is actually there: it runs deep through a chain, where each
    level costs one node, and stops early in a bush. `depth` is still honoured
    when given, so a caller that genuinely wants N levels can ask for them.
    """
    included = {node["name"]}
    frontier = [node]
    level = 0
    while frontier and (depth is None or level < depth):
        nxt = [c for n in frontier for c in (n.get("children") or [])]
        if not nxt:
            break
        if budget is not None and len(included) + len(nxt) > budget:
            break
        included.update(n["name"] for n in nxt)
        frontier = nxt
        level += 1
    return included


def _node_dict(node: dict, included: set[str] | None) -> dict:
    """Serialise `node`, recursing into children present in `included`.

    `included is None` means "this node only" — a stub, marked `truncated` if it
    has children. That is what nodes off a lineage spine get: enough to show the
    label and the size, nothing more.

    `truncated` says "this node has children that are not in this response",
    which is what tells the client whether opening it needs another request or
    is purely a local expand.
    """
    name = node["name"]
    children = node.get("children") or []
    kept = [c for c in children if included is not None and c["name"] in included]

    out = {
        "name": name,
        "rank": node.get("rank", ""),
        "depth": _depth[name],
        "child_count": len(children),
        "species_count": _species_count[name],
        # Total descendants including this node. The UI needs it to say what a
        # full expand would actually render before it commits to rendering it;
        # species_count cannot answer that, since the internal nodes are most of
        # the tree (27k nodes for 18k species).
        "node_count": _node_count[name],
        "truncated": len(kept) < len(children),
        "children": [_node_dict(c, included) for c in kept],
    }
    # Only leaves have these, and only leaves are things you can guess in a game.
    if node.get("common_name"):
        out["common_name"] = node["common_name"]
    if node.get("scientific_name"):
        out["scientific_name"] = node["scientific_name"]
    return out


def subtree(root: str | None = None, depth: int | None = None, budget: int | None = 200) -> dict:
    """Return the subtree at `root`, limited by `depth`, `budget`, or both.

    Raises ValueError for an unknown root, rather than falling back to the tree
    root — silently showing all of Animalia when you asked for Aves is worse
    than an error, because it looks like it worked.
    """
    _ensure_index()
    if root is None:
        node = _tree
    else:
        node = _by_name.get(root)
        if node is None:
            raise ValueError(f"Unknown taxon: {root!r}")
    return _node_dict(node, _select(node, depth, budget))


def path_to(name: str) -> list[str]:
    """Ancestor names from the tree root down to `name`, inclusive.

    Used to jump: the client expands each name in turn, so arriving at a search
    result leaves the whole lineage above it open and readable, rather than
    dropping you into an unmoored subtree.
    """
    _ensure_index()
    if name not in _by_name:
        raise ValueError(f"Unknown taxon: {name!r}")
    chain = [name]
    while chain[-1] in _parent:
        chain.append(_parent[chain[-1]])
    return list(reversed(chain))


# How much of the destination to open when jumping to it. Smaller than a normal
# slice: the spine above already costs nodes, and the point of a jump is to land
# somewhere legible rather than to dump the subtree.
LINEAGE_TARGET_BUDGET = 60


def lineage(name: str) -> dict:
    """The tree from the root down to `name`, with siblings at every level.

    One request rather than one per ancestor. Jumping to a search result by
    fetching each ancestor separately worked but took seventeen round trips for
    a deep species, and left the UI assembling a spine it had no reason to know
    the shape of.

    Nodes off the path come back shallow and `truncated`, so the surrounding
    context is visible without dragging in the rest of the tree. The target
    itself gets a normal slice, so you land on something with children to read
    rather than on a bare label.
    """
    _ensure_index()
    chain = path_to(name)

    def build(node: dict, idx: int) -> dict:
        if idx + 1 >= len(chain):
            return _node_dict(node, _select(node, None, LINEAGE_TARGET_BUDGET))
        nxt = chain[idx + 1]
        out = _node_dict(node, None)
        out["children"] = [
            build(c, idx + 1) if c["name"] == nxt else _node_dict(c, None)
            for c in (node.get("children") or [])
        ]
        out["truncated"] = False
        return out

    return {"path": chain, "tree": build(_tree, 0)}


def search(query: str, limit: int = 25) -> list[dict]:
    """Find taxa and species matching `query`.

    Searches scientific names *and* leaf common names, because in explore mode
    both are things you would reasonably type — the game's `/animals` only
    searches common names, since only a species can be guessed.

    Ordering puts prefix matches first and then larger groups first, so typing
    "cani" offers Canidae before an arbitrary species inside it.
    """
    _ensure_index()
    needle = query.strip().lower()
    if not needle:
        return []

    hits = []
    for name, node in _by_name.items():
        common = node.get("common_name", "")
        hay_sci = name.lower()
        hay_common = common.lower()
        if needle in hay_sci:
            matched = hay_sci
        elif common and needle in hay_common:
            matched = hay_common
        else:
            continue
        hits.append((0 if matched.startswith(needle) else 1, -_species_count[name], name, node, common))

    hits.sort(key=lambda h: (h[0], h[1], h[2]))
    return [
        {
            "name": name,
            "common_name": common,
            "rank": node.get("rank", ""),
            "depth": _depth[name],
            "species_count": _species_count[name],
            "is_species": not (node.get("children") or []),
        }
        for _, _, name, node, common in hits[:limit]
    ]


def stats() -> dict:
    """Size of the loaded tree, so the client can warn before a full expand."""
    _ensure_index()
    return {
        "root": _tree["name"],
        "nodes": _node_count[_tree["name"]],
        "species": _species_count[_tree["name"]],
        "max_depth": max(_depth.values()),
    }

File: frontend/src/engine/explore.ts
/**
 * Free browsing of the tree. Port of `taxoquiz/explore.py`, which explains the
 * node budget, `truncated`, and why a lineage comes back in one piece.
 */
import type { ExploreHit, ExploreNode, ExploreStats, Lineage } from '../api'
import { rankLevels } from './ranks'
import { childrenOf, perTree, pyRepr, type RawNode } from './taxonomy'

interface ExploreIndex {
  tree: RawNode
  byName: Map<string, RawNode>
  parent: Map<string, string>
  depth: Map<string, number>
  speciesCount: Map<string, number>
  nodeCount: Map<string, number>
  warmth: Map<string, number>
}

const exploreIndex = perTree((tree): ExploreIndex => {
  const byName = new Map<string, RawNode>()
  const parent = new Map<string, string>()
  const depth = new Map<string, number>()
  const speciesCount = new Map<string, number>()
  const nodeCount = new Map<string, number>()

  const walk = (node: RawNode, parentName: string | null, d: number): [number, number] => {
    const name = node.name
    byName.set(name, node)
    depth.set(name, d)
    if (parentName !== null) parent.set(name, parentName)
    const children = childrenOf(node)
    if (children.length === 0) {
      speciesCount.set(name, 1)
      nodeCount.set(name, 1)
      return [1, 1]
    }
    let species = 0, nodes = 0
    for (const child of children) {
      const [s, n] = walk(child, name, d + 1)
      species += s
      nodes += n
    }
    speciesCount.set(name, species)
    nodeCount.set(name, nodes + 1)
    return [species, nodes + 1]
  }
  walk(tree, null, 0)
  return { tree, byName, parent, depth, speciesCount, nodeCount, warmth: rankLevels(tree) }
})

/** Names to include, grown breadth-first until a limit bites. `null` is no limit. */
function select(node: RawNode, depth: number | null, budget: number | null): Set<string> {
  const included = new Set([node.name])
  let frontier = [node]
  let level = 0
  while (frontier.length > 0 && (depth === null || level < depth)) {
    const next = frontier.flatMap(childrenOf)
    if (next.length === 0) break
    if (budget !== null && included.size + next.length > budget) break
    for (const n of next) included.add(n.name)
    frontier = next
    level += 1
  }
  return included
}

/** `included === null` serialises the node alone, as a stub. */
function nodeDict(node: RawNode, included: Set<string> | null, idx: ExploreIndex): ExploreNode {
  const name = node.name
  const children = childrenOf(node)
  const kept = included === null ? [] : children.filter((c) => included.has(c.name))
  const out: ExploreNode = {
    name,
    rank: node.rank ?? '',
    depth: idx.depth.get(name)!,
    warmth: idx.warmth.get(name)!,
    child_count: children.length,
    species_count: idx.speciesCount.get(name)!,
    node_count: idx.nodeCount.get(name)!,
    truncated: kept.length < children.length,
    children: kept.map((c) => nodeDict(c, included, idx)),
  }
  if (node.common_name) out.common_name = node.common_name
  if (node.scientific_name) out.scientific_name = node.scientific_name
  return out
}

function unknown(name: string): Error {
  return new Error(`Unknown taxon: ${pyRepr(name)}`)
}

export function subtree(
  tree: RawNode,
  root: string | null = null,
  depth: number | null = null,
  budget: number | null = 200,
): ExploreNode {
  const idx = exploreIndex(tree)
  const node = root === null ? idx.tree : idx.byName.get(root)
  if (node === undefined) throw unknown(root!)
  return nodeDict(node, select(node, depth, budget), idx)
}

export function pathTo(tree: RawNode, name: string): string[] {
  const idx = exploreIndex(tree)
  if (!idx.byName.has(name)) throw unknown(name)
  const chain = [name]
  while (idx.parent.has(chain[chain.length - 1])) chain.push(idx.parent.get(chain[chain.length - 1])!)
  return chain.reverse()
}

export const LINEAGE_TARGET_BUDGET = 60

export function lineage(tree: RawNode, name: string): Lineage {
  const idx = exploreIndex(tree)
  const chain = pathTo(tree, name)
  const build = (node: RawNode, i: number): ExploreNode => {
    if (i + 1 >= chain.length) return nodeDict(node, select(node, null, LINEAGE_TARGET_BUDGET), idx)
    const next = chain[i + 1]
    const out = nodeDict(node, null, idx)
    out.children = childrenOf(node).map((c) => (c.name === next ? build(c, i + 1) : nodeDict(c, null, idx)))
    out.truncated = false
    return out
  }
  return { path: chain, tree: build(idx.tree, 0) }
}

/** Prefix matches first, then larger groups, then by name. */
export function search(tree: RawNode, query: string, limit = 25): ExploreHit[] {
  const idx = exploreIndex(tree)
  const needle = query.trim().toLowerCase()
  if (!needle) return []

  const hits: { prefix: number; size: number; node: RawNode; common: string }[] = []
  for (const [name, node] of idx.byName) {
    const common = node.common_name ?? ''
    const haySci = name.toLowerCase()
    const hayCommon = common.toLowerCase()
    let matched: string
    if (haySci.includes(needle)) matched = haySci
    else if (common && hayCommon.includes(needle)) matched = hayCommon
    else continue
    hits.push({ prefix: matched.startsWith(needle) ? 0 : 1, size: idx.speciesCount.get(name)!, node, common })
  }
  hits.sort((a, b) =>
    a.prefix - b.prefix || b.size - a.size || (a.node.name < b.node.name ? -1 : a.node.name > b.node.name ? 1 : 0))

  return hits.slice(0, limit).map(({ node, common }) => ({
    name: node.name,
    common_name: common,
    rank: node.rank ?? '',
    depth: idx.depth.get(node.name)!,
    species_count: idx.speciesCount.get(node.name)!,
    is_species: childrenOf(node).length === 0,
  }))
}

export function stats(tree: RawNode): ExploreStats {
  const idx = exploreIndex(tree)
  let maxDepth = 0
  for (const d of idx.depth.values()) if (d > maxDepth) maxDepth = d
  return {
    root: idx.tree.name,
    nodes: idx.nodeCount.get(idx.tree.name)!,
    species: idx.speciesCount.get(idx.tree.name)!,
    max_depth: maxDepth,
  }
}
